### PR TITLE
SONARIAC-605 Add clarifying comments to existing and new Docker related APIs

### PR DESCRIPTION
**Pull request by rudy-regazzoni-sonarsource**

### CODE REVIEW

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/AddInstruction.java [hunk 1]
```diff
@@ -19,5 +19,13 @@
...
 package org.sonar.iac.docker.tree.api;
 
+/**
+ * AddInstruction
+ * This interface define the contract of the <a href="https://docs.docker.com/engine/reference/builder/#add">ADD</> instruction.
+ * It is a strict implementation of the {@link TransferInstruction} interface.
+ * <pre>
+ *   {@link #keyword()} {@link #options()} {@link #srcs()} {@link #dest()}
+ * </pre>
+ */
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/AddInstruction.java:29]: I'm wondering if I should in parallel add a literal example, like this :
`ADD --option=value src1 src2 dest`

**Comment by rudy-regazzoni-sonarsource** [comment]
I mostly focused on interface for the public API.
I wrote too much tests at first and then did some clean, as sometime nothing is better than a one line obvious comment.
Please feel free to challenge any comment, I'm pretty sure lot of them are still not worth.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/AddInstruction.java [hunk 2]
```diff
@@ -19,5 +19,13 @@
  */
 package org.sonar.iac.docker.tree.api;
 
+/**
+ * AddInstruction
+ * This interface define the contract of the <a href="https://docs.docker.com/engine/reference/builder/#add">ADD</> instruction.
+ * It is a strict implementation of the {@link TransferInstruction} interface.
+ * <pre>
+ *   {@link #keyword()} {@link #options()} {@link #srcs()} {@link #dest()}
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/AddInstruction.java:27]:
```suggestion
 *   ADD {@link #srcs()} {@link #dest()}
 *   ADD {@link #options()} {@link #srcs()} {@link #dest()}
```

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/AddInstruction.java [hunk 3]
```diff
@@ -19,5 +19,13 @@
  */
 package org.sonar.iac.docker.tree.api;
 
+/**
+ * AddInstruction
+ * This interface define the contract of the <a href="https://docs.docker.com/engine/reference/builder/#add">ADD</> instruction.
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/AddInstruction.java:24]:
```suggestion
 * Interface to define the contract of the <a href="https://docs.docker.com/engine/reference/builder/#add">ADD</> instruction.
```
IMO you don't need `this` all the time. Just use the neutral form.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/Body.java [hunk 4]
```diff
@@ -21,6 +21,15 @@
 
 import java.util.List;
 
+/**
+ * This represents the content of a Dockerfile.
+ * A dockerfile currently contain a list of {@link ArgInstruction} followed by a list of {@link DockerImage}.
+ * The ArgInstruction are considered to be global.
+ * <pre>
+ *   {@link #globalArgs()}
+ *   {@link #dockerImages()}
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/Body.java:30]:
```suggestion
 *   {@link #dockerImages()}
 *   {@link #globalArgs()} {@link #dockerImages()}
```

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/HealthCheckInstruction.java [hunk 5]
```diff
@@ -21,6 +21,17 @@
...
 
+/**
+ * HealthCheckInstruction
+ * This interface define the contract of the <a href="https://docs.docker.com/engine/reference/builder/#healthcheck">HEALTCHECK</> instruction.
+ * It can either be provided with NONE keyword or a {@link CmdInstruction}.
+ * It can also have multiple {@link Flag}.
+ * <pre>
+ *   {@link #keyword()} {@link #none()}
+ *   {@link #keyword()} {@link #cmdInstruction()}
+ *   {@link #keyword()} {@link #options()} {@link #cmdInstruction()}
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/HealthCheckInstruction.java:32]:
```suggestion
 *   HEALTHCHECK {@link #none()}
 *   HEALTHCHECK {@link #cmdInstruction()}
 *   HEALTHCHECK {@link #options()} {@link #cmdInstruction()}
```

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/ShellInstruction.java [hunk 6]
```diff
@@ -19,6 +19,13 @@
  */
 package org.sonar.iac.docker.tree.api;
 
+/**
+ * ShellInstruction
+ * This interface define the contract of the <a href="https://docs.docker.com/engine/reference/builder/#shell">SHELL</> instruction.
+ * <pre>
+ *   {@link #keyword()} {@link #arguments()}
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/ShellInstruction.java:26]:
```suggestion
 *   SHELL {@link #arguments()}
```

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/WorkdirInstruction.java [hunk 7]
```diff
@@ -19,5 +19,12 @@
  */
 package org.sonar.iac.docker.tree.api;
 
+/**
+ * WorkdirInstruction
+ * This interface define the contract of the <a href="https://docs.docker.com/engine/reference/builder/#workdir">WORKDIR</> instruction.
+ * <pre>
+ *   {@link #keyword()} {@link #arguments()}
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/api/WorkdirInstruction.java:26]:
```suggestion
 *   WORKDIR {@link #arguments()}
```

**Review by nils-werner-sonarsource — APPROVED** [review]
It is not easy to find suitable comments for the APIs and classes. I think you have already done a good job.
I have a few comments that you might want to implement:
* A comment does not need a superfluous heading. This means you do not have to write TestUtils if the class is called TestUtils.
* Comments should be contextless. This means that `this` usually refers to a piece of information. (PS you usually forgot the `s` `This Interface defineS`.)
* For instructions where only a special syntax token is allowed as a keyword, you can write the keyword directly.

**Pull request merged**